fix(security): cookie check during GitHub integration installation

oioki opened the pull request:
Adding a special cookie at the beginning of the GitHub integration installation process. At the final step, make sure the cookie is there. This avoids unwanted GitHub installations being attached to a Sentry account. There are two installation flows:

1. Initiated from the Sentry side (via "Add Installation" button on Settings / Organization / Integrations). If the cookie check does not pass, then the following error is displayed:
![image](https://github.com/getsentry/sentry/assets/1127549/c95c0102-8f73-406c-9f54-100a6ba11a98)

2. Initiated from the GitHub side (https://github.com/apps/sentry-io, big "Install" button). If the cookie check fails here, the pop-up window with the same content is shown for a split second, and the installation is denied:
![image](https://github.com/getsentry/sentry/assets/1127549/30e0c5d5-8cbf-444f-98b0-546eca64363d)

The cookie is short-term (10 minutes), hopefully that will be enough to complete the process. In worst case, user can restart the installation process.

oioki committed:
fix(security): cookie check during GitHub integration installation

sentaur-athena reviewed (commented):
I'm not familiar with the security part of the change so will let someone from your team stamp.

oioki commented on src/sentry/web/frontend/pipeline_advancer.py:
```diff
@@ -46,9 +50,17 @@ def handle(self, request: Request, provider_id: str) -> HttpResponseBase:
             and pipeline is None
         ):
```
We did not touch this list for 5 years, so I am going to make it more direct as you said 👍

oioki committed:
add seconds unit

oioki committed:
remove FORWARD_INSTALL_FOR

oioki committed:
add test

mdtro commented on src/sentry/web/frontend/pipeline_advancer.py:
```diff
@@ -40,15 +38,26 @@ def handle(self, request: Request, provider_id: str) -> HttpResponseBase:
+            rv = self.redirect(
                 reverse("integration-installation", args=[provider_id, installation_id])
             )
+            rv.set_cookie(
```
Is there an advantage/requirement here to use `set_cookie` versus the `request.session`? The session is signed so we will be able to prevent tampering.

oioki committed:
save state to session instead

oioki committed:
move document.origin logic to separate method

oioki committed:
Merge branch 'master' into fix/github-integration-csrf

c298lee committed:
chore(feedback): Adds message field to feedback search (#65933)

Allows the message field to appear in feedback search, since the message
is now searchable

schew2381 committed:
fix(api): Improve rate limit query (#67287)

The subquery that we use for fetching the sentry app installation is
slow because we're joining `SentryAppInstallationToken` and
`SentryAppInstallation`.

Instead, speed it up by fetching `SentryAppInstallationToken`
independently first.

Real query info for the `SentryAppInstallationToken` query
```
Index Scan using sentry_sentryappinstallationtoken_api_token_id on sentry_sentryappinstallationtoken (cost=0.29..2.51 rows=1 width=8) (actual time=0.040..0.041 rows=1 loops=1)
	
Index Cond: (api_token_id = 408727)
	
Planning Time: 0.065 ms
	
Execution Time: 0.052 ms
```

---------

Co-authored-by: Mark Story <mark@mark-story.com>

JonasBa committed:
fix(trace): fix txn + span_id scroll (#67316)

@wmak @Zylphrex This fixes the case where span_id is not an actual
standalone span, but in fact points to the transaction. I left the code
comment to explain it.

Commit:
feat(api): allow testing API @deprecated in dev (#67313)

Devserver is self-hosted so it's specifically excluded from marking
endpoints as deprecated.
This change keeps the exclusion for self-hosted, but adds an exception
for devserver.

iamrajjoshi committed:
feat(open-pr-comments): Move PHP to GA (#67298)

Commit:
fix(discord): Fix log messages for Discord (#67304)

Fixes log messages and metric details for Discord such that discord_code
and message can be logged.

`if resp` returns false even when there is a response and mypy doesn't
understand so I added a type ignore.

Abdkhan14 committed:
feat(new-trace): Fixes blank replay viewer in bottom docked drawer. (#67305)

This is for the replay viewer under the transaction tab in the trace
drawer.

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

malwilley committed:
fix(issue-priority): Update copy in dropdown (#67296)

ceorourke committed:
ref(daily summary): Iterate through release projects to find new groups (#67212)

New issues in releases that went out in the last 24 hours are not
populating for the daily summary even though we _do_ have issues first
seen in new releases ([Redash
query](https://redash.getsentry.net/queries/6316/source) that mimics the
django query), but perhaps since it's being limited to the first 2 in
the list it just never happens to have new groups in those. This PR
changes it so it loops over new releases until it finds 2 (or less) that
have groups first seen in them.

Hopefully closes
https://github.com/getsentry/team-core-product-foundations/issues/164

ceorourke committed:
ref(daily summary): Add org slug, format event count, add newline (#67311)

Some more minor notification formatting changes - add the org slug to
the title, format the event count to have commas, add a newline after
the issue title.

Closes
https://github.com/getsentry/team-core-product-foundations/issues/211

narsaynorath committed:
feat(mobile-score): Add mobile profile to relay config (#66812)

Adds the mobile profile that we're treating as an alpha version. This
profile is just to ensure that we're able to ingest the values we want
and we will iterate the profiles in the future.

The browser configs are extracted into a helper function with no
modifications, and there is a mobile profile added. The
`config.performanceScore.profiles` path will only be added if either of
these helpers populates an array with profiles to pass along.

gggritso committed:
feat(perf): Add HTTP module samples panel (#67344)

Just a basic panel for now. Actual contents coming in future PRs. Also,
this introduces some annoying duplication (the headers of these panels
are all the same, but we keep re-creating them) which I'll fix later.

anonrig committed:
ref(typing): add types to nodestore (#67337)

Adds stronger types for nodestore*

Ref: https://github.com/getsentry/team-processing/issues/135

Commit:
ref(crons): Refactor broken detection tests (#67312)

Simplifies some tests here, will be useful as we test more complex cases
with email sending

ykamo001 committed:
debug(integrations): Debug MSTeams integrations (#67317)

"Add to a team" button doesn't seem to be working for the MSTeams bot.
We should verify that the events are coming in as expected from
Microsoft to the bot. We want the `build_team_installation_message`
function/flow to be happening.

cathteng committed:
fix(alert-rule): skip fetching rule snooze owner user if no owner_id (#67350)

evanpurkhiser committed:
style(celery): Formatting consistency on crontab defs (#67318)

Just trying to keep things consistent here

phacops committed:
feat(metrics-indexer): Prevent modifications on hard-coded metric and tag IDs (#67342)

JonasBa committed:
fix(trace): fix crash when activating root node (#67347)

cathteng committed:
chore(integrations): API docs for custom priorities for pagerduty and opsgenie (#67354)

leeandher committed:
fix(event-tags): Highlights section still appears empty (#67284)

Fixes event tags showing up empty
- The reason was because I tried to short circuit the rendering by only
checking `objectIsEmpty(event.contexts)` but it appears there are two
more conditions which could cause `ContextSummary` to render empty.
- Instead of trying to match those to some parent we'll conditionally
show, for now, I will just render the component with the header if it
does end up getting painted (by modifying `contextSummary.tsx`
- Once designs are finalized, this component will not be used any more,
and the internal logic of rendering a minimum amount of contexts, etc.
will go away cleanly.

evanpurkhiser committed:
ref(crons): Split tasks (#67319)

These tasks were starting to get pretty hairy

This splits them into their own modules

scttcper committed:
feat(ui): Remove render function from dropdown label (#67269)

michellewzhang committed:
fix(replays): fix display of selector widgets if no proj selected (#67356)

Fixing a small bug that was hiding the selector widgets if all/no
projects were selected:


https://github.com/getsentry/sentry/assets/56095982/16c7de04-fe14-40ec-959e-a7354bc2bf47

jennmueng committed:
fix(suggested-fix): Send pii_certified to the signal (#67361)

send the pii_certified flag over to the policy check signal

phacops committed:
feat(profiles): Emit profile duration outcomes (#67335)

evanpurkhiser committed:
ref(crons): Allow TIMEOUT complete checkins to have duration updated (#67303)

Currently if an IN_PROGRESS check-in times out there is no way to
determine if a later closing check-in  was sent.

For now we can improve the situation slightly by at least updating the
duration of the timed-out check-in.

This is a partial fix for
https://github.com/getsentry/sentry/issues/54865

aliu39 committed:
perf(replay): parallelize delete_replay_recording with ThreadPoolExecutor (#66881)

- this fx is used by the project_replay_details endpoint DELETE op (used
by customer), and the [delete_replays
script](https://github.com/getsentry/getsentry/pull/13268) (used by OPS)

https://github.com/getsentry/team-replay/issues/15

---------

Co-authored-by: Josh Ferge <josh.ferge@sentry.io>
Co-authored-by: Colton Allen <colton.allen@sentry.io>

Abdkhan14 committed:
feat(new-trace): Limiting displayed issues to 10  (#67357)

Limiting displayed issues to 10 rows and linking to discover to view
entire list:
<img width="1510" alt="Screenshot 2024-03-20 at 1 50 08 PM"
src="https://github.com/getsentry/sentry/assets/60121741/d2f99fc9-b074-454c-bd98-05ae71f34fac">

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

scttcper committed:
fix(settings): Refactor feature hook render function (#67209)

evanpurkhiser committed:
ref(crons): Keep cron tasks colocated and conssitently named (#67359)

Commit:
fix(discord): Fix include_in_slo (#67366)

malwilley committed:
fix(issue-stream): Ensure that bulk action checkboxes are all hidden at the same time (#67352)

This fixes the problem where, at certain viewport widths, it would still
render checkboxes for issues but not in the header.

michellewzhang committed:
ref(replays): rm browser icon, rage/dead click in details for mobile replays (#67355)

For mobile replays, remove the
- browser icon next to URL
- rage/dead click counts at the top

<img width="1248" alt="SCR-20240320-jlss"
src="https://github.com/getsentry/sentry/assets/56095982/5dc69f00-6f04-450e-bac8-b238758a3d08">

web replays still look the same:
<img width="1186" alt="SCR-20240320-jnkt"
src="https://github.com/getsentry/sentry/assets/56095982/612c0fb1-5b03-4efa-bc99-6a0befdd7ecd">

evanpurkhiser committed:
ref(routes): Simplify release routes (#67368)

cathteng committed:
chore(slack): log incoming action payload (#67369)

evanpurkhiser committed:
ref(mypy): A few dict -> Mapping fixes (#67367)

aliu39 committed:
feat(runner): add ingest-feedback-events topic to devserver (#67364)

Adding a new topic to sentry local devservers. This is yet to be tested
in prod, currently feedbacks are still ingested through ingest-events.
See https://github.com/getsentry/sentry/pull/66484

lobsterkatie committed:
fix(grouping): Improve handling of matching grouping configs (#67270)

### Background

In theory, a project's primary and secondary configs should be different, as the secondary config only gets set (to the current config) when there's a new config to which to upgrade (which becomes primary).

However, we've been seeing cases where the values match. Specifically, when we record the hash comparison metric[1], we tag it with both configs. And when we graph that metric, a portion of the occurrences have configs which match. See the PR for a screenshot showing the relative frequency of this happening - it doesn't happen that often, it does happen.

This got me curious, so I added a log[2] to track when this happens. Looking at the logs once they came in, a pattern emerged. A project would have a burst of logging, and then never log again. Also, that burst always seemed to last for around three seconds. (See the PR for a screenshot of this happening.)

I then remembered that while we read the secondary config from project settings, for the primary config we use the value Relay pre-populates in the event[3] and suddenly it made sense. Matt was right[4] - it is a caching issue.

More specifically, the way it happens is that Relay adds the current config (`A`) and sends the event to our ingest pipeline. While in the pipeline, an event's project has its grouping config updated to a newer one (`B`). Now the project's secondary config is `A` (which value we pull) and its current config is `B`. But we continue to use the Relay-populated value (`A`) for the primary config and voila, an event (and a metric) where both configs are `A`.

(We can confirm that this is the cause because we also log when the config value pre-populated by Relay doesn't match the value in project settings[5], and those logs always matched up 1-to-1 with the matching config logs. (See the PR for a screenshot showing this.))

### Changes

To solve this problem, I made a few changes:

- I added a helper function, `_config_update_happened_recently`, which uses the transition period expiry to figure out when the project's config was updated, and returns `True` if that time is within a certain tolerance of the time the function is called, and I used that function to prevent logging if the config was just updated. Despite the 3-second pattern observed in the logs, when using it I gave it a tolerance of 30 seconds, just to be extra conservative.

- I stopped recording the hash comparison metric when the configs are equal.

- I made logging the config equality contingent on the config not having been recently updated.

- Given the link between the Relay/project settings logging and the hash comparison logging, I applied the same logic to the former, so that it also only fires when there's no recent update.


[1] https://github.com/getsentry/sentry/blob/c7e33f856266f23f29ce00a200a762374c165412/src/sentry/grouping/ingest.py#L454-L472
[2] https://github.com/getsentry/sentry/pull/66994
[3] https://github.com/getsentry/sentry/pull/65116
[4] https://github.com/getsentry/sentry/pull/65116#discussion_r1500087757
[5] https://github.com/getsentry/sentry/blob/c1551141c72dbff7a27317b614dcf56150cd08c1/src/sentry/grouping/ingest.py#L261-L278

getsentry-bot committed:
Revert "feat(profiles): Emit profile duration outcomes (#67335)"

This reverts commit b7d24e88e7ec47dfb392fb2dd56086b0ba34cd72.

Co-authored-by: phacops <336345+phacops@users.noreply.github.com>

evanpurkhiser committed:
ref(crons): Hide labels for error / okay (#67372)

This is just extra useless tooltips

ryan953 committed:
ci: Update frontend file-filters to be each tailored to the tools they relate to (#67251)

Updates the file-filters lists to be more specific for each tool. There
are 3 tool categories: linting js (eslint and biome), linting css
(stylelint), testing files.

Note that each of these 3 categories includes a list of files to check,
but also a list of configs that should trigger a run too. So if someone
adds a new lint rule to the config we'll re-run everything right away.

anonrig committed:
chore: add more tracing to processing path (#67378)

Add more traces to  `save_event_transaction` paths

Ref: https://github.com/getsentry/team-processing/issues/135

ceorourke committed:
ref(daily summary): Filter to unresolved issues (#67265)

Update `project_key_errors` to filter out unresolved issues. This wasn't
possible until recently, but now we can join the `group_attributes`
entity which has `group_status` on it in snuba. Since the weekly reports
and daily summaries use this function, fixing this kills 2 birds. I put
this behind a feature flag to be cautious and not take down snuba.

Closes
https://github.com/getsentry/team-core-product-foundations/issues/209

malwilley committed:
fix(issue-stream): Invalidate cached data modifying issues from stream (#67362)

This fixes the following problem with stale issue details data:

1. Hover over an issue in the stream, which loads it into the cache
2. Change priority or resolve/archive that issue from the stream
3. Click into that issue and see that it still shows the old state

phacops committed:
feat(profiles): Emit profile duration outcomes (#67380)

ceorourke committed:
ref(daily summary): Replace pytz usage with zoneinfo (#67381)

I guess we don't use pytz anymore

JonasBa committed:
fix(trace): dont render replays (#67376)

Dont render replays if we dont have a replayId

scttcper committed:
test(ui): Add missing await to waitFors (#67390)

scttcper committed:
test(ui): Wait for waitFor in members test (#67392)

anonrig committed:
chore: add more tracing around event manager (#67389)

ykamo001 committed:
debug(msteams): needs more logging (#67391)

Needs more logging, as I'm not seeing enough events. It's starting to
look like we're not handling the webhook events properly from MSTeams

JonasBa committed:
fix(trace): fix react 18 false positive (#67400)

Fix false positive error

michellewzhang committed:
ref(replays): move start time in replay details (#67397)

Move the start time to be under the user info:

<img width="1268" alt="SCR-20240320-nwos"
src="https://github.com/getsentry/sentry/assets/56095982/4224b986-7c33-49af-bbf1-71cb5901aebe">

Based on Figma:
https://www.figma.com/file/bbTpXNRRu7NY4MDaAAFiJe/Specs%3A-Session-Replay-Mobile-Alpha-1?type=design&node-id=1-13024&mode=design&t=QKMYY86lUUQvBNcH-0

malwilley committed:
fix(alerts): Refetch configs when changing the alert project (#67403)

Alerts can return different sets of actions/conditions/rules depending
on the selected project. One example is the priority alert condition
which is enabled in certain projects with a feature flag. However,
currently the UI doesn't refetch the new config data after the project
has changed.

This is a quick fix for that which refetches the configuration in the
background. It already does this for environments, so I added this in a
similar fashion.

gggritso committed:
fix(perf): Improve `Chart` with `isBarChart` behaviour (#67363)

Passing `isBarChart` to the `Chart` component tries to render a bar
chart, naturally. Right now, no UIs make use of `isBarChart` but I want
to! I can't though, 'cause a bunch of things about it are a little
broken. I think this code path was left in the dust while we improved
the other chart types.

In short, I added some specific props to make the bar chart behave more
sensibly. It needs different options from line and area charts.

- better X and Y axis labels and behaviour
- simpler tooltips
- removed properties that don't apply

Probably there's more work to be done, but this is a start.

antonpirker committed:
Do not use deprecated Sentry SDK features (#67417)

Makes some tiny changed to not use deprecated Sentry SDK features and
thus making the code more future proof.

obostjancic committed:
ref(dashboards): align dashboard metrics types (#67348)

oioki committed:
another attempt to fix
